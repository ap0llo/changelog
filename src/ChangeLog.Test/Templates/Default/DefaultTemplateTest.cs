using Grynwald.ChangeLog.Configuration;
using Grynwald.ChangeLog.Model;
using Grynwald.ChangeLog.Model.Text;
using Grynwald.ChangeLog.Templates;
using Grynwald.ChangeLog.Templates.Default;
using Xunit;

namespace Grynwald.ChangeLog.Test.Templates.Default
{
    /// <summary>
    /// Tests for <see cref="DefaultTemplate"/>
    /// </summary>
    public class DefaultTemplateTest : ScribanBaseTemplateTest
    {
        protected override ITemplate GetTemplateInstance(ChangeLogConfiguration configuration) => new DefaultTemplate(configuration);

        protected override void SetCustomDirectory(ChangeLogConfiguration configuration, string customDirectory)
        {
            configuration.Template.Default.CustomDirectory = customDirectory;
        }



        [Fact]
        public void Name_returns_expected_value()
        {
            // ARRANGE
            var sut = GetTemplateInstance(ChangeLogConfigurationLoader.GetDefaultConfiguration());

            // ACT 
            var actual = sut.Name;

            // ASSERT
            Assert.Equal(TemplateName.Default, actual);
        }

        [Fact]
        public void ChangeLog_with_multiple_versions_is_converted_to_expected_Output()
        {
            // Changelog with two versions.
            // All versions are empty (no entries)

            var changeLog = new ApplicationChangeLog()
            {
                GetSingleVersionChangeLog("1.2.3"),
                GetSingleVersionChangeLog("4.5.6"),
            };

            Approve(changeLog);
        }


        private class CustomTextElement : INormalizedTextElement
        {
            public string NormalizedText { get; set; } = "";

            public TextStyle NormalizedStyle { get; set; }

            public string Text { get; set; } = "";

            public TextStyle Style { get; set; }
        }

        [Fact]
        public void When_normalization_is_enabled_normalized_text_is_rendered()
        {
            // ARRANGE
            var testData = new TestDataFactory();
            var configuration = new ChangeLogConfiguration();
            configuration.Template.Default.NormalizeReferences = true;

            var footerValue = new CustomTextElement()
            {
                Text = "Text",
                Style = TextStyle.Code,
                NormalizedText = "NormalizedText",
                NormalizedStyle = TextStyle.None
            };

            var changeLog = new ApplicationChangeLog()
            {
                testData.GetSingleVersionChangeLog("1.2.3", entries: new[]
                {
                    testData.GetChangeLogEntry(footers: new[]
                    {
                        new ChangeLogEntryFooter(
                            new("Name"),
                            footerValue
                        )
                    })
                })
            };

            // ACT / ASSERT
            Approve(changeLog, configuration);
        }

        [Fact]
        public void When_normalization_is_disabled_default_text_is_rendered()
        {
            // ARRANGE
            var testData = new TestDataFactory();
            var configuration = new ChangeLogConfiguration();
            configuration.Template.Default.NormalizeReferences = false;

            var footerValue = new CustomTextElement()
            {
                Text = "Text",
                Style = TextStyle.Code,
                NormalizedText = "NormalizedText",
                NormalizedStyle = TextStyle.None
            };

            var changeLog = new ApplicationChangeLog()
            {
                testData.GetSingleVersionChangeLog("1.2.3", entries: new[]
                {
                    testData.GetChangeLogEntry(footers: new[]
                    {
                        new ChangeLogEntryFooter(
                            new("Name"),
                            footerValue
                        )
                    })
                })
            };

            // ACT / ASSERT
            Approve(changeLog, configuration);
        }


    }
}
