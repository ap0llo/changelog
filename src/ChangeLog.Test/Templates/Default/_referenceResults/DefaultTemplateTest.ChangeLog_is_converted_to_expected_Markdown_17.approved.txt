# Change Log

## 1.2.3

#### <a id="changelog-heading-0003E8"></a> **scope1:** Some change

- see\-also: [Link](http://example.com/)
- Commit: `0003E8`
